src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface ILspServiceProvider
+{
+    ILspServices CreateServices(string serverKind);
+}
+
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+}
+
+public interface IRequestExecutionQueue<RequestContextType> where RequestContextType : struct
+{
+    Task<RequestContextType?> CreateRequestContextAsync(IQueueItem<RequestContextType> queueItem, CancellationToken cancellationToken);
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
+
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(
+        bool mutatesSolutionState,
+        bool requiresLSPSolution,
+        IRequestHandler<TRequestType, TResponseType, RequestContextType> handler,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
```

Comment:
Would it be valuable to construct handlers with this provided instead of per-message? Or does it not matter?

Reply:
Yeah I think it's not the handler's job to know any state outside what's directly related to its job (like a response cache).

In both Roslyn and Razor I'm currently planning to include the ClientCapabilities on the RequestContext.